src/hash-join/join_compute_api.h

```diff
@@ -44,166 +49,181 @@ struct join_pair { size_type first, second; };
+  build_hash_table<<<build_grid_size, block_size>>>(hash_table.get(), 
+                                                    build_column, 
+                                                    build_column_length);
+  
+  CUDA_RT_CALL( cudaGetLastError() );
+
+  // Allocate storage for the counter used to get the size of the join output
+  size_type * d_join_output_size;
+  cudaMalloc(&d_join_output_size, sizeof(size_type));
+  cudaMemset(d_join_output_size, 0, sizeof(size_type));
+
+  // Probe with the left table
+  gdf_table<size_type> const & probe_table{left_table};
+  const size_type probe_column_length{probe_table.get_column_length()};
+  const key_type * const probe_column{static_cast<key_type*>(probe_table.get_probe_column_data())};
+  const size_type probe_grid_size{(probe_column_length + block_size -1)/block_size};
+
+  // Probe the hash table without actually building the output to simply
+  // find what the size of the output will be.
+  compute_join_output_size<join_type, 
+                           multimap_type, 
+                           key_type, 
+                           size_type,
+                           block_size, 
+                           DEFAULT_CUDA_CACHE_SIZE>
+	<<<probe_grid_size, block_size>>>(hash_table.get(), 
+                                    build_table, 
+                                    probe_table, 
+                                    probe_column,
+                                    probe_table.get_column_length(),
+                                    d_join_output_size);
+
+  CUDA_RT_CALL( cudaGetLastError() );
+
+  // Copy output size from the device to host
+  size_type h_join_output_size{0};
+  CUDA_RT_CALL( cudaMemcpy(&h_join_output_size, d_join_output_size, sizeof(size_type), cudaMemcpyDeviceToHost));
+  
+  // If the output size is zero, return immediately
+  if(0 == h_join_output_size){
+    return error;
   }
 
-  CUDA_RT_CALL( cudaMallocManaged   ( &tempOut, sizeof(joined_type)*scanSize));
-  CUDA_RT_CALL( cudaMemPrefetchAsync( tempOut , sizeof(joined_type)*scanSize, dev_ordinal));
-
-  CUDA_RT_CALL( cudaMemset(d_joined_idx, 0, sizeof(size_type)) );
-  // step 3b: scan table A (left), probe the HT and output the joined indices - doing left join here
-  probe_hash_tbl<join_type, multimap_type, key_type, key_type2, key_type3, size_type, joined_type, block_size, DEFAULT_CUDA_CACHE_SIZE>
-	<<<(a_count + block_size-1) / block_size, block_size>>>
-	(hash_tbl.get(), a, a_count, a2, b2, a3, b3,
-	 static_cast<joined_type*>(tempOut), d_joined_idx, scanSize);
-  error = cudaDeviceSynchronize();
+  // Allocate modern GPU storage for the join output
+  int dev_ordinal{0};
+  join_output_pair* tempOut{nullptr};
+  CUDA_RT_CALL( cudaGetDevice(&dev_ordinal));
+  joined_output = mgpu::mem_t<size_type> (2 * (h_join_output_size), compute_ctx);
+
+  // Allocate temporary device buffer for join output
+  CUDA_RT_CALL( cudaMallocManaged   ( &tempOut, sizeof(join_output_pair)*h_join_output_size));
```

Comment:
Why are we writing to a temporary uvm buffer instead of directly to the gdf_columns that will eventually represent the result?